docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,543 @@
+spec:
+  targetRef:
+    kind: Mesh
+  to:
+    - targetRef:
+        kind: MeshService
+        name: backend
+      default: $conf1
+```
+
+The result for `backend` will be `merge($conf1, $conf2)` because `aaa < bbb`.
+While intuitively the conf shouldn't depend on the policy name in this case and be `merge($conf2, $conf1)`,
+because `kind: MeshService` is more specific than `kind: Mesh` and we clearly have to prioritize `conf1`.
+
+Now is a good moment to improve the way sorting/merging works, because ResourceRules is a new structure
+and it's going to be used only when computing config for the new MeshServices.
+
+#### New approach
+
+We're going to take all matched policies and produce a list of the following items:
+
+```yaml
+toItemsWithTopLevel:
+  - topLevelTargetRefKind: Mesh
+    policyName: aaa
+    toTargetRefKind: Mesh
+    conf: $conf2
+  - topLevelTargetRefKind: Mesh
+    policyName: bbb
+    toTargetRefKind: MeshService
+    conf: $conf1
+```
+
+and we'll sort this list based on fields in the following order:
+
+1. topLevelTargetRefKind
+2. toTargetRefKind
+3. policyName
+
+Note: today we basically do 1 and 3, ignoring the field 2.
+
+After that, if we want to find a conf for the resource `r`, we do:
+
+```go
+confs := []
+for _, item := range toItemsWithTopLevel:
+    if includes(item, r) {
+        confs = append(confs, item.conf)
+    }
+}
+resultingConf := merge(confs...)
+```
+
+Function `includes(r1, r2 core_model.Resource) bool` checks if one resource is included into another:
+* every resource is included into `Mesh`
+* `Mesh*Route` is included into `MeshService` if route's `to[].targetRef` references the MeshService
+
+#### Migration
+
+ResourceRule is a new structure, it will be used when configuring outbounds for the real MeshService resources.
+That's why we don't need backwards compatibility for `to[]`.
+
+For `from[]` policies the sorting algorithm should be changed as well. 
+But the problem is less visible due to the fact `from[].targetRef.kind` is `Mesh` in the overwhelming majority of cases.
+For MeshTrafficPermission it can be `Mesh` and `MeshSubset`, and technically the behaviour of the following policies is going to change:
+
+```yaml
+kind: MeshTrafficPermission
+name: aaa
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: Mesh
+      default: 
+        action: Deny
+---
+kind: MeshTrafficPermission
+name: bbb
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshSubset
+        tags:
+          kuma.io/service: backend
+      default: 
+        action: Allow
+```
+
+Current behaviour is 'action: Deny' for all, because `aaa` happens to be more specific than `bbb`.
+The new behaviour is 'action: Deny' for all except `backend` and 'action: Allow' for `backend`.
+If such change in behaviour happens it probably means the user already had policies in the obscure state. 
+The change is harmless, and we don't need to provide a migration path for `from[]`.
```

Comment:
I would say we need to document this. The current behaviour just seems a bit broken. Might make sense to think about any "sane" cases where this change could break something for the end user.

Reply:
Yeah, there is actually one kind of "sane" case, when mesh operator decided to step in and override the behaviour configured by the application developers. I'll update the MADR with this info